source/server/listener_manager_impl.cc

```diff
@@ -281,111 +292,215 @@ bool ListenerImpl::isWildcardServerName(const std::string& name) {
+    addFilterChainForServerNames(destination_ips_map[EMPTY_STRING], server_names,
+                                 transport_protocol, application_protocols, filter_chain);
+  } else {
+    for (const auto& destination_ip : destination_ips) {
+      addFilterChainForServerNames(destination_ips_map[destination_ip], server_names,
+                                   transport_protocol, application_protocols, filter_chain);
+    }
+  }
+}
+
+void ListenerImpl::addFilterChainForServerNames(
+    ServerNamesMap& server_names_map, const std::vector<std::string>& server_names,
+    const std::string& transport_protocol, const std::vector<std::string>& application_protocols,
+    const Network::FilterChainSharedPtr& filter_chain) {
   if (server_names.empty()) {
-    addFilterChainForApplicationProtocols(filter_chains_[EMPTY_STRING][transport_protocol],
+    addFilterChainForApplicationProtocols(server_names_map[EMPTY_STRING][transport_protocol],
                                           application_protocols, filter_chain);
   } else {
     for (const auto& server_name : server_names) {
       if (isWildcardServerName(server_name)) {
         // Add mapping for the wildcard domain, i.e. ".example.com" for "*.example.com".
         addFilterChainForApplicationProtocols(
-            filter_chains_[server_name.substr(1)][transport_protocol], application_protocols,
+            server_names_map[server_name.substr(1)][transport_protocol], application_protocols,
             filter_chain);
       } else {
-        addFilterChainForApplicationProtocols(filter_chains_[server_name][transport_protocol],
+        addFilterChainForApplicationProtocols(server_names_map[server_name][transport_protocol],
                                               application_protocols, filter_chain);
       }
     }
   }
 }
 
 void ListenerImpl::addFilterChainForApplicationProtocols(
-    std::unordered_map<std::string, Network::FilterChainSharedPtr>& transport_protocol_map,
+    ApplicationProtocolsMap& application_protocols_map,
     const std::vector<std::string>& application_protocols,
     const Network::FilterChainSharedPtr& filter_chain) {
   if (application_protocols.empty()) {
-    transport_protocol_map[EMPTY_STRING] = filter_chain;
+    application_protocols_map[EMPTY_STRING] = filter_chain;
   } else {
     for (const auto& application_protocol : application_protocols) {
-      transport_protocol_map[application_protocol] = filter_chain;
+      application_protocols_map[application_protocol] = filter_chain;
+    }
+  }
+}
+
+void ListenerImpl::finishFilterChain() {
+  for (auto& port : filter_chains_) {
+    auto& destination_ips_pair = port.second;
+    auto& destination_ips_map = destination_ips_pair.first;
+    std::vector<std::pair<ServerNamesMapSharedPtr, std::vector<Network::Address::CidrRange>>> list;
+    for (const auto& entry : destination_ips_map) {
+      std::vector<Network::Address::CidrRange> subnets;
+      if (entry.first == EMPTY_STRING) {
```

Comment:
entry.first.empty() ?

Reply:
Done, although I feel that the old check was a bit more correct, since we're checking if it's the `map[EMPTY_STRING]` entry.

Reply:
Oh, right, I forgot that's what it is doing. In that case, I'd be ok with putting it back to how you had it.